tests/utils/node/nodes.go

```diff
@@ -0,0 +1,177 @@
+	runtimeErrors = make([]<-chan error, 0, len(nodes))
+	for _, node := range nodes {
+		runtimeError, err := node.Start(ctx)
+		if err != nil {
+			return runtimeErrors, fmt.Errorf("node with index %d: %w",
+				*node.index, err)
+		}
+
+		runtimeErrors = append(runtimeErrors, runtimeError)
+	}
+
+	for _, node := range nodes {
+		port := node.RPCPort()
+		err := waitForNode(ctx, port)
+		if err != nil {
+			return runtimeErrors, fmt.Errorf("node with index %d: %w", *node.index, err)
+		}
+	}
+
+	return runtimeErrors, nil
+}
+
+// InitAndStartTest is a test helper method to initialise and start nodes,
+// as well as registering appriopriate test handlers.
+// If any node fails to initialise or start, cleanup is done and the test
+// is instantly failed.
+// If any node crashes at runtime, all other nodes are shutdown,
+// cleanup is done and the passed argument `signalTestToStop`
+// is called to signal to the main test goroutine to stop.
+func (nodes Nodes) InitAndStartTest(ctx context.Context, t *testing.T,
+	signalTestToStop context.CancelFunc) {
+	t.Helper()
+
+	err := nodes.Init(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	nodesCtx, nodesCancel := context.WithCancel(ctx)
+	runtimeErrors := runtime.NewErrorsFanIn()
+
+	for _, node := range nodes {
+		runtimeError, err := node.Start(nodesCtx) // takes little time
+		if err == nil {
+			runtimeErrors.Add(node.String(), runtimeError)
+			continue
+		}
+
+		t.Errorf("Node %s failed to start: %s", node, err)
+
+		stopNodes(t, nodesCancel, runtimeErrors)
+		t.FailNow()
+	}
+
+	// this is run sequentially since all nodes start almost at the same time
+	// so waiting for one node will also wait for all the others.
+	// You can see this since the test logs out that all the nodes are ready
+	// at the same time.
+	for _, node := range nodes {
+		err := waitForNode(ctx, node.RPCPort())
+		if err == nil {
+			t.Logf("Node %s is ready", node)
+			continue
+		}
+
+		t.Errorf("Node %s failed to be ready: %s", node, err)
+		stopNodes(t, nodesCancel, runtimeErrors)
+		t.FailNow()
+	}
+
+	// watch for runtime fatal error from any of the nodes
+	watchDogCtx, watchDogCancel := context.WithCancel(ctx)
+	watchDogDone := make(chan struct{})
+	go func() {
+		defer close(watchDogDone)
+		err := runtimeErrors.Watch(watchDogCtx)
+		watchDogWasStopped := errors.Is(err, context.Canceled) ||
+			errors.Is(err, context.DeadlineExceeded)
+		if watchDogWasStopped {
+			return
+		}
+
+		t.Errorf("one node has crashed: %s", err)
+		// we cannot stop the test with t.FailNow() from a goroutine
+		// other than the test goroutine, so we call failNow to signal
+		// it to the test goroutine.
+		signalTestToStop()
+	}()
+
+	t.Cleanup(func() {
+		t.Helper()
+		// Disengage node watchdog goroutine
+		watchDogCancel()
+		<-watchDogDone
+		// Stop and wait for nodes to exit
+		stopNodes(t, nodesCancel, runtimeErrors)
+	})
+}
+
+func stopNodes(t *testing.T, nodesCancel context.CancelFunc,
```

Comment:
curious as to why this isn't just a receiver function on `Nodes`?

Reply:
Because:

- all the arguments for this function are not part of the `Nodes`
- the calling tests do not need to stop multiple nodes so far

So I don't think there is a need to have it as a method. If one day a test needs to stop multiple nodes, then yeah we should move it as a receiver.